pkg/cmd/service/service.go

```diff
@@ -21,68 +19,191 @@ import (
+	HTTP       HTTPConfig
+	Authn      AuthnConfig
+	Log        LogConfig
+	Playground PlaygroundConfig
+}
 
-	// Authentication. Possible options: none,preshared,oidc
-	AuthMethod string `default:"none" split_words:"true"`
+// DefaultConfig returns the OpenFGA server default configurations.
+func DefaultConfig() Config {
+	return Config{
+		OpenFGA: OpenFGAConfig{
+			MaxTuplesPerWrite:             100,
+			MaxTypesPerAuthorizationModel: 100,
+			ChangelogHorizonOffset:        0,
+			ResolveNodeLimit:              25,
+		},
+		Database: DatabaseConfig{
+			Engine:       "memory",
+			MaxCacheSize: 100000,
+		},
+		GRPC: GRPCConfig{
+			Enabled: true,
+			Addr:    ":8081",
+			TLS:     TLSConfig{Enabled: false},
+		},
+		HTTP: HTTPConfig{
+			Enabled:                true,
+			Addr:                   ":8080",
+			TLS:                    TLSConfig{Enabled: false},
+			UpstreamRequestTimeout: 3 * time.Second,
+			CORSAllowedOrigins:     []string{"*"},
+			CORSAllowedHeaders:     []string{"*"},
+		},
+		Authn: AuthnConfig{
+			Method: "none",
+		},
+		Log: LogConfig{
+			Format: "text",
+		},
+		Playground: PlaygroundConfig{
+			Enabled: false,
+			Port:    3000,
+		},
+	}
+}
 
-	// Shared key authentication
-	AuthPresharedKeys []string `default:"" split_words:"true"`
+// GetServiceConfig returns the OpenFGA server configuration based on the values provided in the server's 'config.yaml' file.
+// The 'config.yaml' file is loaded from '/etc/openfga', '$HOME/.openfga', or the current working directory. If no configuration
+// file is present, the default values are returned.
+func GetServiceConfig() (Config, error) {
 
-	// OIDC authentication
-	AuthOIDCIssuer   string `default:"" split_words:"true"`
-	AuthOIDCAudience string `default:"" split_words:"true"`
+	config := DefaultConfig()
 
-	// Logging. Possible options: text,json
-	LogFormat string `default:"text" split_words:"true"`
+	viper.SetConfigName("config")
+	viper.SetConfigType("yaml")
 
-	PlaygroundEnabled bool `default:"true" split_words:"true"`
-	PlaygroundPort    int  `default:"3000" split_words:"true"`
+	configPaths := []string{"/etc/openfga", "$HOME/.openfga", "."}
+	for _, path := range configPaths {
+		viper.AddConfigPath(path)
+	}
+	viper.AutomaticEnv()
 
-	// CORS configuration
-	CORSAllowedOrigins []string `default:"*" split_words:"true"`
-	CORSAllowedHeaders []string `default:"*" split_words:"true"`
-}
+	err := viper.ReadInConfig()
+	if err != nil {
+		_, ok := err.(viper.ConfigFileNotFoundError)
+		if ok {
+			// if the server config is not found then return the defaults
+			return config, nil
+		}
 
-func GetServiceConfig() Config {
-	var config Config
+		return config, fmt.Errorf("failed to load server config: %w", err)
+	}
 
-	if err := envconfig.Process("OPENFGA", &config); err != nil {
-		log.Fatalf("failed to process server config: %v", err)
+	if err := viper.Unmarshal(&config); err != nil {
+		return config, fmt.Errorf("failed to unmarshal server config: %w", err)
```

Comment:
Should we return an empty config if we also return an error?

Reply:
I usually like to return nil values if an error is encountered, but the signature is currently `(Config, error)`.

I can change it to return a `Config{}, ...`, but what do you think about changing the signature to a pointer return value and just returning `nil`?

Reply:
That works for me 👍

Reply:
Ok, I made that change 👍